_posts/2017-09-20-mvc-delievery-mechanism-dm.md

```diff
@@ -0,0 +1,178 @@
+</center>
+<br/> 
+
+Although reducing boiler plate code using frameworks sounds a great idea, we are coupling the structure of our entities (domain model) to the APIs we provide to the external world and our database. Changing the database impacts the API and vice-versa. What had initially been seen as a time saver, now is a reason to avoid change. The larger and more coupled the system grows, the less developers will be inclined to make changes due to the size of the ripple effects. Many web apps today have APIs which are far from ideal as they reflect the internal data structure of the application instead of focusing only on the information that would make more sense to the external world. 
+
+Do not be afraid to create your own mappers. They are normally very simple to write and there are many small libraries specialised to parse objects to JSON or persist data to databases. The advantage of writing our own mappers is that we do not need to couple our APIs or databases to anything. Changes are localised and easy to change. On top of that, we can easily test-drive our mappers and move API tests to the unit level instead of doing it at Acceptance level. 
+
+## Deployment options
+
+The biggest advantage of keeping the Domain Model decoupled from the Delivery Mechanism is that it give us deployment options as the application evolves. 
+
+While the application is small, we can keep the separation between delivery mechanism and domain model just using good package/namespace structure or sub-modules, but still keep everything in the same project and deploy both together, as a single application. 
+
+**Embedded Domain Model**
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-09-20-mvc-delivery-mechanism/embedded-dm.png" alt=“Embedded Domain Model” class="img img-responsive" style="height: 70%; width: 70%;"/>
+</center>
+<br/> 
+
+As the application grows and we want to add different delivery mechanisms, or use different technology stacks, or scale delivery mechanism and domain model in different ways, it makes sense to deploy them individually. If the separation is already there, it would not be so hard to achieve that. We just need to wrap the Domain Model in some infrastructure and change the invocation from the Delivery Mechanism.
+
+**Deployable Domain Model**
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-09-20-mvc-delivery-mechanism/deployable-dm.png" alt="Deployable Domain Model" class="img img-responsive" style="height: 70%; width: 70%;"/>
+</center>
+<br/> 
+
+# Summary
+
+In MVC, Controllers should be a very thin layer (2 or 3 lines) which identify the request from the user, invoke the appropriate behaviour in the Model, and invoke the appropriate View once the Model returns. Model is not only about entities (state); it’s about all the behaviour of your system, the entire Domain Model.
```

Comment:
Disagree with the method size (in line numbers). 

Even if all your business logic is handled in services/model, you need to deal with the conversion from internal (business) errors/messages to HTTP errors + statuses. 

Personally, I'd keep it as 

> very thin layer (to the absolute minimum) which identify

...

Reply:
Makes sense. the number of lines is not the main point here and will take the focus away from the actual message. I'll change it.